### PR TITLE
add SOCKS upstream proxy support (socks4/socks5)

### PR DESCRIPTION
**Pull request by rofl0r**
original patch submitted in 2006 to debian mailing list:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=392848%29#12

this version was rebased to git and updated by Russ Dill <russ.dill@gmail.com>
in 2015 (the original patch used a different config file format).

as discussed in #40.

commit message by @rofl0r.

### CODE REVIEW

File: src/reqs.c [hunk 1]
```diff
@@ -1265,6 +1267,88 @@ static void relay_connection (struct conn_s *connptr)
         return;
 }
 
+static int
+connect_to_upstream_proxy(struct conn_s *connptr, struct request_s *request)
+{
+	int len;
```
Comment (rofl0r) [inline src/reqs.c]: len needs to be unsigned.

File: src/reqs.c [hunk 2]
```diff
@@ -1265,6 +1267,88 @@ static void relay_connection (struct conn_s *connptr)
...
+	char buff[512]; /* won't use more than 7 + 255 */
```
Comment (rofl0r) [inline src/reqs.c]: this needs to be unsigned char, because ...

File: src/reqs.c [hunk 3]
```diff
@@ -1265,6 +1267,88 @@ static void relay_connection (struct conn_s *connptr)
...
+			return -1;
+		if (buff[0]!=5 || buff[1]!=0)
+			return -1;
+		switch(buff[3]) {
+			case 1: len=4; break; /* ip v4 */
+			case 4: len=16; break; /* ip v6 */
+			case 3: /* domainname */
+				if (1 != safe_read(connptr->server_fd, buff, 1))
+					return -1;
+				len = buff[0]; /* max = 255 */
```
Comment (rofl0r) [inline src/reqs.c]: since buff is currently signed char, the len will be set to a negative number if it's bigger than 127.

**Comment by obnoxxx** [comment]
Can the original author be contacted?
Because I am not sure we should add his code to the offical repo with his authorship without his concent...

**Comment by rofl0r** [comment]
the commit has his email address, so yes. but i suppose he sent it to the debian mailing list in the hope that it would be used, not shelfed. also since tinyproxy is GPL licensed, all patched made to it are GPL as well and if the author shared his code there's nothing speaking against merging it.
the only thing that *I* think could be sensible is that his email is getting into the commit history (spammer alert) but then it was already publicly shared on the debian bugtracker as well, so...

**Comment by obnoxxx** [comment]
I sent a mail. If we don't hear from him, we can still add the patch (possibly with @rofl0r or s/o else as author and mentioning the original author in the commit message).

**Comment by crosscutrat** [comment]
Thank you @rofl0r !  Is there a way to donate to the project?  This one will be a great help!

**Comment by rofl0r** [comment]
iirc @obnoxxx is not interested, but you could support my work via my other projects [proxychains-ng](https://github.com/rofl0r/proxychains-ng) or [sabotage-linux](http://github.com/sabotage-linux/sabotage) with bitcoin (address is mentioned at the bottom of the READMEs of the respective projects)

**Comment by obnoxxx** [comment]
@rofl0r, I have not heard back from the author. I guess I will prepare a patch that mentions the history, but not with their authorship...

@crosscutrat, @rofl0r, never said I'm not interested in donations... :-D ,  but there is currently no way of donating to the project as such. Individuals could take donations. I might consider bitcoin.  I could offer paypal.

**Comment by TampertK** [comment]
What's the status of this request? When can it be merged?

**Comment by fanthos** [comment]
Is there some updates for this PR?
Currently I need to use Privoxy for HTTP to SOCKS proxy.

**Comment by rofl0r** [comment]
@obnoxxx i've added the fixes from my review, improved documentation and tested. imo it's good to merge. please review.

**Comment by chromer030** [comment]
Hi , Does this tool supports Socks protocol ? , if yes please mention it's support in Bio.

**Comment by rofl0r** [comment]
> Hi , Does this tool supports Socks protocol ? , if yes please mention it's support in Bio.

bio? the tool supports socks upstream proxies, but it isn't a socks proxy itself.

**Comment by chromer030** [comment]
I mean Description , This tool is great , thanks for keeping it up to date.

**Pull request merged**